File: practice/alphazero/game_board.py
#!/usr/bin/env python3
"""
C++ GameBoard with Python fallback
"""

import sys
import os
import numpy as np

try:
    from fast_game_board import GameBoard as CPPGameBoard
    
    class GameBoard(CPPGameBoard):
        """C++ GameBoard with Python compatibility"""
        
        def get_score(self):
            result = super().get_score()
            return (result[0], result[1])  # tuple 인덱스 접근
        
        def get_winner(self):
            winner = super().get_winner()
            return winner if winner != -1 else None
        
        
        def get_state_tensor(self, perspective_player):
            state_list = super().get_state_tensor(perspective_player)
            return np.array(state_list, dtype=np.float32)
        
        # Python과 호환성을 위한 추가 메서드들 (pybind11에서 property는 이미 정의됨)
        def is_terminal(self):
            return self.is_game_over()
    
    print("C++ GameBoard 사용")
    
except ImportError:
    # Python 폴백
    print("C++ GameBoard 빌드 실패, Python 구현 사용 (성능 저하 예상)")
    
    from typing import List, Tuple, Optional
    import copy

    class GameBoard:
        """
        버섯 게임 보드 클래스
        - 10x17 보드에서 합이 10인 직사각형을 선택하는 게임
        - 플레이어 0과 1이 번갈아가며 플레이
        """
        
        def __init__(self, initial_board: List[List[int]]):
            self.R = 10  # 행 수
            self.C = 17  # 열 수
            self.board = copy.deepcopy(initial_board)
            self.current_player = 0  # 0: 첫 번째 플레이어, 1: 두 번째 플레이어
            self.pass_count = 0  # 연속 패스 횟수
            self.game_over = False
            self.winner = None
            
        
        def copy(self):
            """게임 보드 복사본 생성"""
            new_board = GameBoard([[0] * self.C for _ in range(self.R)])
            new_board.board = copy.deepcopy(self.board)
            new_board.current_player = self.current_player
            new_board.pass_count = self.pass_count
            new_board.game_over = self.game_over
            new_board.winner = self.winner
            return new_board
        
        def get_valid_moves(self) -> List[Tuple[int, int, int, int]]:
            """현재 상태에서 유효한 움직임 반환 (조기 종료 최적화 적용)"""
            if self.game_over:
                return []
            
            valid_moves = []
            
            for r1 in range(self.R):
                for c1 in range(self.C):
                    skip_larger_r2 = False
                    for r2 in range(r1, self.R):
                        if skip_larger_r2:
                            break
                        for c2 in range(c1, self.C):
                            # 면적 체크
                            area = (r2 - r1 + 1) * (c2 - c1 + 1)
                            if area < 2:
                                continue
                            
                            # 합계 계산
                            total_sum = self._get_box_sum(r1, c1, r2, c2)
                            
                            if total_sum >= 10:
                                if total_sum == 10 and self._check_edges(r1, c1, r2, c2):
                                    valid_moves.append((r1, c1, r2, c2))
                                # 같은 r2에서 더 큰 c2들은 건너뛰기
                                break
                            
                            # 세로 한 줄(c1==c2)에서 합>=10이면 더 큰 r2들도 건너뛰기
                            if c1 == c2 and total_sum >= 10:
                                skip_larger_r2 = True
            
            return valid_moves
        
        def _get_box_sum(self, r1: int, c1: int, r2: int, c2: int) -> int:
            """박스 내부 점수 합계 계산 (양수만)"""
            total_sum = 0
            for i in range(r1, r2 + 1):
                for j in range(c1, c2 + 1):
                    if self.board[i][j] > 0:
                        total_sum += self.board[i][j]
            return total_sum
        
        def _check_edges(self, r1: int, c1: int, r2: int, c2: int) -> bool:
            """네 변에 각각 최소 하나 이상의 버섯이 있는지 확인"""
            top, down, left, right = False, False, False, False
            
            # 상단과 하단 변
            for j in range(c1, c2 + 1):
                if self.board[r1][j] > 0:
                    top = True
                if self.board[r2][j] > 0:
                    down = True
            
            # 좌측과 우측 변
            for i in range(r1, r2 + 1):
                if self.board[i][c1] > 0:
                    left = True
                if self.board[i][c2] > 0:
                    right = True
            
            return top and down and left and right
        
        def _is_valid_move(self, r1: int, c1: int, r2: int, c2: int) -> bool:
            """움직임이 유효한지 검사"""
            # 범위 체크
            if not (0 <= r1 <= r2 < self.R and 0 <= c1 <= c2 < self.C):
                return False
            
            area = (r2 - r1 + 1) * (c2 - c1 + 1)
            if area < 2:  # 최소 2칸 이상
                return False
            
            # 합이 10인지 확인
            total_sum = self._get_box_sum(r1, c1, r2, c2)
            if total_sum != 10:
                return False
            
            # 네 변에 각각 최소 하나 이상의 버섯이 있는지 확인
            return self._check_edges(r1, c1, r2, c2)
        
        def make_move(self, r1: int, c1: int, r2: int, c2: int, player: int) -> bool:
            """움직임 실행"""
            if self.game_over:
                return False
            
            # 패스인 경우
            if r1 == -1 and c1 == -1 and r2 == -1 and c2 == -1:
                self.pass_count += 1
                if self.pass_count >= 2:
                    self._end_game()
                else:
                    self.current_player = 1 - self.current_player
                return True
            
            # 유효한 움직임인지 확인
            if not self._is_valid_move(r1, c1, r2, c2):
                return False
            
            # 영역 점령
            for i in range(r1, r2 + 1):
                for j in range(c1, c2 + 1):
                    self.board[i][j] = -(player + 1)  # -1은 플레이어 0, -2는 플레이어 1
            
            self.pass_count = 0  # 패스 카운트 초기화
            self.current_player = 1 - self.current_player
            
            # 더 이상 유효한 움직임이 없으면 게임 종료
            if not self.get_valid_moves():
                self._end_game()
            
            return True
        
        def _end_game(self):
            """게임 종료 처리"""
            self.game_over = True
            
            # 점수 계산
            score = [0, 0]
            for row in self.board:
                for cell in row:
                    if cell == -1:
                        score[0] += 1
                    elif cell == -2:
                        score[1] += 1
            
            # 승자 결정
            if score[0] > score[1]:
                self.winner = 0
            elif score[1] > score[0]:
                self.winner = 1
            else:
                self.winner = -1  # 무승부
        
        def is_terminal(self) -> bool:
            """게임이 끝났는지 확인"""
            return self.game_over
        
        def get_winner(self) -> Optional[int]:
            """승자 반환 (0, 1, -1(무승부), None(게임 진행 중))"""
            if self.game_over:
                return self.winner
            return None
        
        def get_reward(self, player: int) -> float:
            """플레이어의 보상 반환"""
            if not self.game_over:
                return 0.0
            
            if self.winner == player:
                return 1.0
            elif self.winner == -1:  # 무승부
                return 0.0
            else:
                return -1.0
        
        def get_state_tensor(self, perspective_player: int) -> np.ndarray:
            """신경망 입력용 상태 텐서 생성 (2, 10, 17)"""
            state = np.zeros((2, self.R, self.C), dtype=np.float32)
            
            for i in range(self.R):
                for j in range(self.C):
                    cell = self.board[i][j]
                    if cell > 0:
                        # 버섯 값을 정규화 (1-9 -> 0.1-0.9)
                        state[0][i][j] = cell / 10.0
                    elif cell == -(perspective_player + 1):
                        # 현재 플레이어가 점령한 칸
                        state[1][i][j] = 1.0
                    elif cell == -(2 - perspective_player):
                        # 상대 플레이어가 점령한 칸
                        state[1][i][j] = -1.0
            
            return state
        
        def get_score(self) -> Tuple[int, int]:
            """현재 점수 반환 (플레이어 0 점수, 플레이어 1 점수)"""
            score = [0, 0]
            for row in self.board:
                for cell in row:
                    if cell == -1:
                        score[0] += 1
                    elif cell == -2:
                        score[1] += 1
            return tuple(score)
        
        def display(self) -> str:
            """보드 상태를 문자열로 표현"""
            result = []
            result.append(f"Current Player: {self.current_player}")
            result.append(f"Pass Count: {self.pass_count}")
            result.append(f"Game Over: {self.game_over}")
            if self.game_over:
                result.append(f"Winner: {self.winner}")
            
            score = self.get_score()
            result.append(f"Score - Player 0: {score[0]}, Player 1: {score[1]}")
            result.append("")
            
            # 보드 출력
            for i, row in enumerate(self.board):
                row_str = ""
                for j, cell in enumerate(row):
                    if cell > 0:
                        row_str += f"{cell:2d} "
                    elif cell == -1:
                        row_str += "P0 "
                    elif cell == -2:
                        row_str += "P1 "
                    else:
                        row_str += " . "
                result.append(f"{i:2d}: {row_str}")
            
            return "\n".join(result)
        
        def __str__(self) -> str:
            return self.display()

File: practice/alphazero/cpp/src/game_board.cpp
#include "game_board.h"
#include <algorithm>
#include <iostream>

GameBoard::GameBoard(const std::vector<std::vector<int>>& initial_board)
    : board(initial_board), current_player(0), pass_count(0), game_over(false), winner(-1) {
    build_action_mapping();
}

GameBoard::GameBoard(const GameBoard& other)
    : board(other.board), current_player(other.current_player), 
      pass_count(other.pass_count), game_over(other.game_over), 
      winner(other.winner), action_to_move(other.action_to_move),
      move_to_action(other.move_to_action) {
}

void GameBoard::build_action_mapping() {
    int action_idx = 0;
    
    // 모든 가능한 직사각형에 대해 매핑 생성
    for (int r1 = 0; r1 < R; ++r1) {
        for (int c1 = 0; c1 < C; ++c1) {
            for (int r2 = r1; r2 < R; ++r2) {
                for (int c2 = c1; c2 < C; ++c2) {
                    int area = (r2 - r1 + 1) * (c2 - c1 + 1);
                    if (area >= 2) {  // 최소 2칸 이상
                        Move move = std::make_tuple(r1, c1, r2, c2);
                        action_to_move[action_idx] = move;
                        move_to_action[move_to_key(r1, c1, r2, c2)] = action_idx;
                        action_idx++;
                    }
                }
            }
        }
    }
    
    // 패스 액션 추가
    Move pass_move = std::make_tuple(-1, -1, -1, -1);
    action_to_move[action_idx] = pass_move;
    move_to_action[move_to_key(-1, -1, -1, -1)] = action_idx;
}

long long GameBoard::move_to_key(int r1, int c1, int r2, int c2) const {
    // 음수 처리를 위해 offset 추가
    long long key = ((long long)(r1 + 100) << 24) | 
                    ((long long)(c1 + 100) << 16) | 
                    ((long long)(r2 + 100) << 8) | 
                    (c2 + 100);
    return key;
}

std::vector<GameBoard::Move> GameBoard::get_valid_moves() const {
    if (game_over) {
        return {};
    }
    
    std::vector<Move> valid_moves;
    
    for (int r1 = 0; r1 < R; ++r1) {
        for (int c1 = 0; c1 < C; ++c1) {
            bool skip_larger_r2 = false;
            for (int r2 = r1; r2 < R; ++r2) {
                if (skip_larger_r2) {
                    break;
                }
                for (int c2 = c1; c2 < C; ++c2) {
                    // 면적 체크
                    int area = (r2 - r1 + 1) * (c2 - c1 + 1);
                    if (area < 2) {
                        continue;
                    }
                    
                    // 합계 계산
                    int total_sum = get_box_sum(r1, c1, r2, c2);
                    
                    if (total_sum >= 10) {
                        if (total_sum == 10 && check_edges(r1, c1, r2, c2)) {
                            valid_moves.emplace_back(r1, c1, r2, c2);
                        }
                        // 같은 r2에서 더 큰 c2들은 건너뛰기
                        break;
                    }
                    
                    // 세로 한 줄(c1==c2)에서 합>=10이면 더 큰 r2들도 건너뛰기
                    if (c1 == c2 && total_sum >= 10) {
                        skip_larger_r2 = true;
                    }
                }
            }
        }
    }
    
    return valid_moves;
}

int GameBoard::get_box_sum(int r1, int c1, int r2, int c2) const {
    int total_sum = 0;
    for (int i = r1; i <= r2; ++i) {
        for (int j = c1; j <= c2; ++j) {
            if (board[i][j] > 0) {
                total_sum += board[i][j];
            }
        }
    }
    return total_sum;
}

bool GameBoard::check_edges(int r1, int c1, int r2, int c2) const {
    bool top = false, down = false, left = false, right = false;
    
    // 상단과 하단 변
    for (int j = c1; j <= c2; ++j) {
        if (board[r1][j] > 0) top = true;
        if (board[r2][j] > 0) down = true;
    }
    
    // 좌측과 우측 변
    for (int i = r1; i <= r2; ++i) {
        if (board[i][c1] > 0) left = true;
        if (board[i][c2] > 0) right = true;
    }
    
    return top && down && left && right;
}

bool GameBoard::is_valid_move(int r1, int c1, int r2, int c2) const {
    // 범위 체크
    if (!(0 <= r1 && r1 <= r2 && r2 < R && 0 <= c1 && c1 <= c2 && c2 < C)) {
        return false;
    }
    
    int area = (r2 - r1 + 1) * (c2 - c1 + 1);
    if (area < 2) {  // 최소 2칸 이상
        return false;
    }
    
    // 합이 10인지 확인
    int total_sum = get_box_sum(r1, c1, r2, c2);
    if (total_sum != 10) {
        return false;
    }
    
    // 네 변에 각각 최소 하나 이상의 버섯이 있는지 확인
    return check_edges(r1, c1, r2, c2);
}

bool GameBoard::make_move(int r1, int c1, int r2, int c2, int player) {
    if (game_over) {
        return false;
    }
    
    // 패스인 경우
    if (r1 == -1 && c1 == -1 && r2 == -1 && c2 == -1) {
        pass_count++;
        if (pass_count >= 2) {
            end_game();
        } else {
            current_player = 1 - current_player;
        }
        return true;
    }
    
    // 유효한 움직임인지 확인
    if (!is_valid_move(r1, c1, r2, c2)) {
        return false;
    }
    
    // 영역 점령
    for (int i = r1; i <= r2; ++i) {
        for (int j = c1; j <= c2; ++j) {
            board[i][j] = -(player + 1);  // -1은 플레이어 0, -2는 플레이어 1
        }
    }
    
    pass_count = 0;  // 패스 카운트 초기화
    current_player = 1 - current_player;
    
    // 더 이상 유효한 움직임이 없으면 게임 종료
    if (get_valid_moves().empty()) {
        end_game();
    }
    
    return true;
}

void GameBoard::end_game() {
    game_over = true;
    
    // 점수 계산
    int score[2] = {0, 0};
    for (const auto& row : board) {
        for (int cell : row) {
            if (cell == -1) {
                score[0]++;
            } else if (cell == -2) {
                score[1]++;
            }
        }
    }
    
    // 승자 결정
    if (score[0] > score[1]) {
        winner = 0;
    } else if (score[1] > score[0]) {
        winner = 1;
    } else {
        winner = -1;  // 무승부
    }
}

bool GameBoard::is_terminal() const {
    return game_over;
}

std::pair<int, int> GameBoard::get_score() const {
    int score[2] = {0, 0};
    for (const auto& row : board) {
        for (int cell : row) {
            if (cell == -1) {
                score[0]++;
            } else if (cell == -2) {
                score[1]++;
            }
        }
    }
    return {score[0], score[1]};
}

GameBoard GameBoard::copy() const {
    return GameBoard(*this);
}

int GameBoard::get_action_space_size() const {
    return action_to_move.size();
}

int GameBoard::encode_move(int r1, int c1, int r2, int c2) const {
    auto it = move_to_action.find(move_to_key(r1, c1, r2, c2));
    if (it != move_to_action.end()) {
        return it->second;
    }
    return -1;  // 찾을 수 없음
}

GameBoard::Move GameBoard::decode_action(int action_idx) const {
    auto it = action_to_move.find(action_idx);
    if (it != action_to_move.end()) {
        return it->second;
    }
    return std::make_tuple(-1, -1, -1, -1);  // 기본값
}

std::vector<std::vector<std::vector<float>>> GameBoard::get_state_tensor(int perspective_player) const {
    std::vector<std::vector<std::vector<float>>> state(2, 
        std::vector<std::vector<float>>(R, std::vector<float>(C, 0.0f)));
    
    for (int i = 0; i < R; ++i) {
        for (int j = 0; j < C; ++j) {
            int cell = board[i][j];
            if (cell > 0) {
                // 버섯 값을 정규화 (1-9 -> 0.1-0.9)
                state[0][i][j] = cell / 10.0f;
            } else if (cell == -(perspective_player + 1)) {
                // 현재 플레이어가 점령한 칸
                state[1][i][j] = 1.0f;
            } else if (cell == -(2 - perspective_player)) {
                // 상대 플레이어가 점령한 칸
                state[1][i][j] = -1.0f;
            }
        }
    }
    
    return state;
}

float GameBoard::get_reward(int player) const {
    if (!game_over) {
        return 0.0f;
    }
    
    if (winner == player) {
        return 1.0f;
    } else if (winner == -1) {  // 무승부
        return 0.0f;
    } else {
        return -1.0f;
    }
}